agreement_rebate/models/agreement_rebate_settlement.py

```diff
@@ -0,0 +1,362 @@
+
+    @api.depends(
+        "invoice_line_ids",
+        "invoice_line_ids.parent_state",
+        "invoice_line_ids.refund_line_ids",
+    )
+    def _compute_invoice_status(self):
+        for line in self:
+            if line.invoice_status == "no":
+                continue
+            invoice_lines = line.invoice_line_ids.filtered(
+                lambda ln: ln.parent_state != "cancel"
+            )
+            refund_lines = invoice_lines.refund_line_ids.filtered(
+                lambda ln: ln.parent_state != "cancel"
+            )
+            if invoice_lines and not refund_lines:
+                line.invoice_status = "invoiced"
+            else:
+                line.invoice_status = "to_invoice"
+
+    def write(self, vals):
+        res = super().write(vals)
+        if "active" in vals and not self.env.context.get(
+            "skip_active_field_update", False
+        ):
+            if vals["active"]:
+                # If one line is active settlement must be active
+                settlements = self.mapped("settlement_id").filtered(
+                    lambda s: not s.active
+                )
+            else:
+                # If lines are archived and the settlement has not active lines, the
+                # settlement must be archived
+                settlements = self.mapped("settlement_id").filtered(
+                    lambda s: s.active and not s.line_ids
+                )
+            settlements.with_context(skip_active_field_update=True).active = vals[
+                "active"
+            ]
+        return res
+
+    def _prepare_invoice(self):
+        """
+        Prepare the dict of values to create the new invoice for a sales order.
+        This method may be overridden to implement custom invoice generation
+        (making sure to call super() to establish a clean extension chain).
+        """
+        self.ensure_one()
+        company = self.company_id or self.env.user.company_id
+        partner = self.env.context.get("partner_invoice", False)
+        if not partner:
+            invoice_group = self.env.context.get("invoice_group", "settlement")
+            if invoice_group == "settlement":
+                partner = self.settlement_id.partner_id
+            elif invoice_group == "partner":
+                partner = self.partner_id
+            elif invoice_group == "commercial_partner":
+                partner = self.partner_id.commercial_partner_id
+        invoice_type = self.env.context.get("invoice_type", "out_invoice")
+        journal_id = (
+            self.env.context.get("journal_id")
+            or self.env["account.move"]
+            .with_company(company=company)
+            .default_get(["journal_id"])["journal_id"]
+        )
+        if not journal_id:
+            raise UserError(
+                _("Please define an accounting sales journal for" " this company.")
+            )
+        vinvoice = self.env["account.move"].new(
```

Comment:
In this version, now you can avoid this and directly use the original dictionary. See for example:

https://github.com/odoo/odoo/blob/2af74c260a6223256e3a975f1378bfa25728edfb/addons/account/tests/common.py#L674

Reply:
Sorry, I'm not getting your point, on V15 same function exists

https://github.com/odoo/odoo/blob/15.0/addons/account/tests/common.py#L644

Reply:
Yes, but it was not complete and fill all the data. Now they are thanks to the computed writable fields.

Reply:
Check a similar conversion in https://github.com/OCA/l10n-spain/commit/3b661814aa91feb7a6fe385079d9010fee4c9acd